[BB-3622] Add course creation condition for organisation

farhaanbukhsh opened the pull request:
## Product Change Description


#### Current State

Studio, as of today doesn't have a way to restrict a user to create a course in a particular organization. What Studio provides right now is a `CourseCreator` permission which gives an `Admin` the power to grant a user the permission to create a course.

For example: If the  `Admin` has given a user `Spiderman` the permission to create courses, `Spiderman` can now create courses in any organization i.e `Marvel` as well as `DC`.

There is no way to restrict `Spiderman` from creating courses under `DC`.

#### Purpose of this PR

The changes done here gives `Admin` the ability to restrict a user on an Organization level from creating courses,

For example: Now, the `Admin` can give the user `Spiderman` the privilege of creating courses only under `Marvel` organization. The moment `Spiderman` tries to create a course under some other organization(i.e DC), we are showing the following error message:

![image](https://user-images.githubusercontent.com/7670449/128211264-ff9e3265-c462-4653-a687-c922c831e409.png)


###### Admin's Responsibility

With this PR the `Admin` can give users the permission to create courses under `any` organization or under a `specific` set of organizations. 

If `Admin` checks the `All Organization` checkbox then the `user` will be granted permission to create courses under any organization.

![Selection_024](https://user-images.githubusercontent.com/7670449/130602326-a2971e7a-610f-4138-a0d4-1595a9f592ee.png)

If `Admin` adds `specific` organizations under the selected lists then the `user` will only be able to create courses under those organizations.

![Selection_025](https://user-images.githubusercontent.com/7670449/130603248-9dc81093-b912-4c2d-9dec-c66de81355fd.png)


**There are few edge cases that can be encountered and below are the screenshot of how we inform the Admin/User about it**


**When `All Organization` is unchecked and no `Organization` is added to the list**

![image](https://user-images.githubusercontent.com/7670449/128209710-11b91724-e973-400c-9eb6-4488fbd25559.png)

**When `All Organization` is checked and specific `Organization` are added as well**

![image](https://user-images.githubusercontent.com/7670449/128210135-a13128ad-a60e-4763-b925-93db9e562956.png)

**Error message when the user enters organization name for which user doesn't have permission**

![image](https://user-images.githubusercontent.com/7670449/128211264-ff9e3265-c462-4653-a687-c922c831e409.png)




## Engineering Implementation

The condition added to ensure that if the feature is enabled
user will not be able to create the course outside of the organisation
in which they belong.

**JIRA tickets**: [BB-3622](https://tasks.opencraft.com/browse/BB-3622) 

~~**Discussions**: Link to any public dicussions about this PR or the design/architecture. Otherwise omit this.~~

**Dependencies**: None

**Screenshots**: 
![image](https://user-images.githubusercontent.com/7670449/108390150-61d3c100-7236-11eb-914c-efb1b8781425.png)



~~**Sandbox URL**: TBD - sandbox is being provisioned.~~

~~**Merge deadline**: "None" if there's no rush, "ASAP" if it's critical, or provide a specific date if there is one.~~

**Testing instructions**:

1. Get the master devstack up and running
2. Pull this branch in edx-platform
3. Make sure to run the migration
3. Drop in the studio shell `make dev.shell.studio`
4. vim /edx/etc/studio.yml and set ENABLE_CREATOR_GROUP = True
The below changes are also required to be done
```diff
diff --git a/cms/envs/devstack.py b/cms/envs/devstack.py
index 9d610b31f0..b548167a35 100644
--- a/cms/envs/devstack.py
+++ b/cms/envs/devstack.py
@@ -152,7 +152,7 @@ CREDENTIALS_SERVICE_USERNAME = 'credentials_worker'
 FEATURES['CERTIFICATES_HTML_VIEW'] = True
 
 ########################## AUTHOR PERMISSION #######################
-FEATURES['ENABLE_CREATOR_GROUP'] = False
+FEATURES['ENABLE_CREATOR_GROUP'] = True
 
 ################### FRONTEND APPLICATION PUBLISHER URL ###################
 FEATURES['FRONTEND_APP_PUBLISHER_URL'] = 'http://localhost:18400'
diff --git a/cms/envs/devstack_decentralized.py b/cms/envs/devstack_decentralized.py
index f7d61d41a1..f115c9e56a 100644
--- a/cms/envs/devstack_decentralized.py
+++ b/cms/envs/devstack_decentralized.py
@@ -107,7 +107,7 @@ CREDENTIALS_SERVICE_USERNAME = 'credentials_worker'
 FEATURES['CERTIFICATES_HTML_VIEW'] = True
 
 ########################## AUTHOR PERMISSION #######################
-FEATURES['ENABLE_CREATOR_GROUP'] = False
+FEATURES['ENABLE_CREATOR_GROUP'] = True
 
 ################### FRONTEND APPLICATION PUBLISHER URL ###################
 FEATURES['FRONTEND_APP_PUBLISHER_URL'] = 'http://localhost:18400'
```

5. ~~In the `FEATURES` dictionary add  `RESTRICT_NON_ORG_COURSE_CREATION: true`~~
6. Now create a user
7. ~~Go to the `admin` page and in localhost:18010/admin/student/courseaccessrole/ create a new  entry, add the user, give it instructor privilege, don't add the `Org` yet~~
8. Log in the studio and request for permission to create the course
9. Log in with a user who is a staff and superuser in another browser(edx on devstack)
10. Go to [http://localhost:18010/admin/course_creators/coursecreator/](http://localhost:18010/admin/course_creators/coursecreator/)
11. Here you can see the permission, all_organization and organization field
10. If you enable all_organzitaions field it will create a CourseCreator Role and works accordingly(The old flow)
11. If you disable all_organization field then you need to add Organizations, this will create OrganiztionContentCreator Role for that organization
12. Then you can try creating a course and in the Org field you need to enter org short name(Same as the one you gave permission to create in the previous role), this is case sensitive field and it will create the course
13. If any other Org is entered, it will not create the course and will through error
14. Another area that can be tested is on enabling `all_organization field`, course_content_group role s created in [http://localhost:18010/admin/student/courseaccessrole/](http://localhost:18010/admin/student/courseaccessrole/) and on disabling it and adding org, OrgContentCreator role is created.

Additional Manual Tests

1. In the course creator admin, change the status to granted and check if roles are created
2. Change the status to anything but granted and check if roles gets removed, roles can be checked in [http://localhost:18010/admin/student/courseaccessrole/](http://localhost:18010/admin/student/courseaccessrole/)

**Author notes and concerns**:

1.  It's a in discovery feature
2. The update to m2m field is non atomic operation, hence https://timonweb.com/django/many-to-many-field-save-method-and-the-django-admin-caveat/ was taken in consideration to obtain the desired result.

**Reviewers**
- [ ] @lgp171188 
- [ ] @xitij2000

edx-webhook commented:
Thanks for the pull request, @farhaanbukhsh! I've created [OSPR-5631](https://openedx.atlassian.net/browse/OSPR-5631) to keep track of it in JIRA, where we prioritize reviews. Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


This is currently a draft pull request. When it is ready for our review and all tests are green, click "Ready for Review", or remove "WIP" from the title, as appropriate.

<!-- comment:end_of_wip -->




<!-- comment:external_pr -->
<!-- jenkins ok to test -->
<!-- data: eyJkcmFmdCI6IHRydWV9 -->

natabene commented:
@farhaanbukhsh Thank you for your contribution. Please let me know once it is ready for our review.

lgp171188 reviewed (approved):
:+1: Can you remove the `[WIP]` tag from the PR and mark it ready by moving it out of the draft status?

- [x] I tested the changes in this PR when reviewing https://github.com/open-craft/edx-platform/pull/319 and verified that:
  * Superusers and global staff users can always create courses in any organization with or without this feature flag enabled.
  * When the feature flag is enabled, in addition to having the course creation permission via course creator groups (enabled by default), the user should also have a role (any role would suffice as the user is already a course creator) in the course's organization to be able to create a course under an organization.
  * When the feature flag is disabled (the current and default behavior), the existing behavior allowing course creator users to create courses in all organizations is preserved.
- [x] I read through the code
- [x] ~~I checked for accessibility issues~~ NA
- [x] ~~Includes documentation~~ NA

natabene commented:
@farhaanbukhsh Is this ready for edX review?

farhaanbukhsh commented:
@natabene this PR is ready for review. :) Thanks for your patience.

kdmccormick commented:
Thanks for your patience @farhaanbukhsh , I'll take a look at this soon!

kdmccormick reviewed (changes requested):
First off, I think this feature is a great addition to CMS. It seems like a no-brainer that we should be able to restrict content creation based on org. Thanks for submitting this PR!

My concern is that we are adding it in a way that is orthogonal to the current content creation auth scheme. As far as I can tell, the current scheme is essentially:

* if `DISABLE_{COURSE,LIBRARY}_CREATION`, reject all content creation, except for global staff.
* if `ENABLE_CREATOR_GROUP`, check if the user has the `CourseCreatorRole` role.
* Otherwise, allow all content creation.

But by building the org-based creation restriction using `has_studio_write_access`, we are instead checking the Instructor and Staff roles, which seem to be intended for already-created content. The resulting flow is that a complicated interaction between several different roles would determine one's effective creation permissions.

I believe it would be better to instead build this on top of the existing `CourseCreatorRole` and the related access checking functions. What do you think?

farhaanbukhsh commented:
Hey @kdmccormick, thanks a lot for reviewing the code. I agree with the fact that the flow is a bit complicated. I am just wondering what can be a better approach?

As I see it we check [CourseCreator](https://github.com/edx/edx-platform/blob/41e5403f4ec4bfa6b1f53036903e6d3318f062e8/cms/djangoapps/contentstore/views/course.py#L852) permission here.

So we introduce a check here, if the user has a CourseCreator role then it goes ahead and does a check that the user should be present in an organization using the courseaccessrole model.

Am I understanding this correct? Thanks a lot for the help here again :)

kdmccormick commented:
@farhaanbukhsh Right, so there is a [CMS model called CourseCreator](https://github.com/edx/edx-platform/blob/master/cms/djangoapps/course_creators/models.py) that tracks status of a user's request to be able to create content in Studio. I could see us adding a nullable `organization` column, foreign-keying to [organizations.models.Organization](https://github.com/edx/edx-organizations/blob/master/organizations/models.py).

That model raises a signal which is caught in [course_creators/views.py](https://github.com/edx/edx-platform/blob/master/cms/djangoapps/course_creators/views.py), which updates the `CourseCreatorRole`. You could modify these functions so that if `organization` is defined, you instead update a newly-defined `OrgCourseCreatorRole` (or, better yet, `OrgContentCreatorRole`, since this role will allow library creation as well).

That `CourseCreatorRole` is defined in [student/roles.py](https://github.com/edx/edx-platform/blob/master/common/djangoapps/student/roles.py) as a subclass of `RoleBase`. Similarly, you could define `OrgCourseCreatorRole` there, much like `OrgStaff` is defined. 

Finally, in `contentstore/views/{course,library}`, you could update the permissions checks to look at both `CourseCreatorRole` (for global permissions) and `OrgCourseCreatorRole` (for org-specific permissions). As a bonus, you don't even need a new feature flag for this -- you can just use the existing `COURSE_CREATOR_GROUP` flag, since this change should be backwards-compaible.

farhaanbukhsh commented:
@kdmccormick  Thanks for the elaborate explanation, this looks like a legitimate good solution. I will try to implement this and ping once done :smile:

farhaanbukhsh commented:
@kdmccormick

> Right, so there is a CMS model called CourseCreator that tracks status of a user's request to be able to create content in Studio. I could see us adding a nullable organization column, foreign-keying to organizations.models.Organization.

A [Many to many field](https://docs.djangoproject.com/en/3.1/topics/db/examples/many_to_many/) makes more sense here, what do you say?

kdmccormick commented:
@farhaanbukhsh 

> A Many to many field makes more sense here, what do you say? Since a course creator can be a part of many organizations.

Good point! How do you suggest we indicate that a CourseCreator has global creation access, though? Treating an empty many-to-many relation as "global access" seems dicey; it would be very surprising if removing every org from the relation gave you access to create courses under *every* org!

Perhaps: a boolean `all_organizations` field, which we could backfill as `True`, so that existing entries continue to be treated as global access.

farhaanbukhsh commented:
> Perhaps: a boolean `all_organizations` field, which we could backfill as `True`, so that existing entries continue to be treated as global access.

This looks like a good approach for now where we can use the boolean, to toggle between CourseCreator Role and OrgContentCreator Role.

So if the boolean is False, we don't create CoureCreatorRole for the user instead for each Organization we give the user an OrgContentCreator role. If the boolean is switched we just need to add/remove ContentCreatorRole accordingly.

Hope I am understanding this correctly. I am working on this and will push the changes soon. :)

kdmccormick commented:
@farhaanbukhsh Sounds right to me!

lgp171188 commented on cms/djangoapps/course_creators/migrations/0002_auto_20210418_0806.py:
```diff
@@ -0,0 +1,24 @@
+# Generated by Django 2.2.20 on 2021-04-18 08:06
```
@farhaanbukhsh, nit: please use a proper name for the migration file instead of the auto-generated one.

lgp171188 commented on cms/djangoapps/contentstore/views/course.py:
```diff
@@ -849,16 +850,18 @@ def _create_or_rerun_course(request):
+                             auth.user_has_role(request.user, OrgContentCreatorRole(org=org)))
+
+        if not is_course_creator:
+            raise PermissionDenied()
```
@farhaanbukhsh, this will show the `Network error` to the user in the Studio UI and that is not very clear. It will be better to have a clear error message when a user cannot create a course under an organization if he/she doesn't have the course creation privilege for that organization.

lgp171188 commented on cms/djangoapps/course_creators/admin.py:
```diff
@@ -30,26 +33,53 @@ def get_email(obj):
+        add_role = updated_state == CourseCreator.GRANTED
+        if not add_role and orgs:
+            raise ValidationError("Organizations cannot be added if the role is not granted.")
+        if orgs and all_orgs:
+            raise ValidationError("All Organization should be disabled to add organization")
+        if not orgs and not all_orgs:
+            raise ValidationError("Organizations should be added if All Organization is disabled.")
```
@farhaanbukhsh, maybe these two could be combined into a single `if` condition and a combined error message like.

``The role can be granted either to all organizations or to specific organizations but not both.```
?

lgp171188 commented on cms/djangoapps/course_creators/models.py:
```diff
@@ -48,6 +48,11 @@ class CourseCreator(models.Model):
+    orgs = models.ManyToManyField(Organization, blank=True,
+                                  help_text=_("Organizations for which content creator is a part of"))
+    all_organizations = models.BooleanField(default=True,
+                                            help_text=_("To restrict the course creation by user in \
+                                                specific organizations"))
```
@farhaanbukhsh, it is cleaner to use the implicit string concatenation instead of escaping the new-line character with a `\`. I have also suggested a different help text which starts with an action verb.
```suggestion
                                            help_text=_("Grant the user the permission to create courses "                                                                                                                 
                                                        "in all organizations"))
```

lgp171188 commented on cms/djangoapps/course_creators/models.py:
```diff
@@ -48,6 +48,11 @@ class CourseCreator(models.Model):
     note = models.CharField(max_length=512, blank=True, help_text=_("Optional notes about this user (for example, "
                                                                     "why course creation access was denied)"))
+    orgs = models.ManyToManyField(Organization, blank=True,
+                                  help_text=_("Organizations for which content creator is a part of"))
```
```suggestion
                                  help_text=_("Organizations under which the course creator is allowed "                                                                                                                   
                                              "to create courses"))
```

lgp171188 commented on cms/djangoapps/course_creators/admin.py:
```diff
@@ -141,3 +173,17 @@ def send_admin_notification_callback(sender, **kwargs):  # lint-amnesty, pylint:
+
+
+@receiver(m2m_changed, sender=CourseCreator.orgs.through)
+def post_all_organizations_callback(sender, **kwargs):  # lint-amnesty, pylint: disable=unused-argument
```
```suggestion
def course_creator_organizations_changed_callback(sender, **kwargs):  # lint-amnesty, pylint: disable=unused-argument
```

lgp171188 commented on cms/djangoapps/course_creators/admin.py:
```diff
@@ -30,26 +33,53 @@ def get_email(obj):
+        all_orgs = self.cleaned_data.get("all_organizations")
+        updated_state = self.cleaned_data.get("state")
+        orgs = self.cleaned_data.get("orgs").exists()
+        add_role = updated_state == CourseCreator.GRANTED
+        if not add_role and orgs:
+            raise ValidationError("Organizations cannot be added if the role is not granted.")
```
@farhaanbukhsh, this causes issues when a request is initially granted for a specific set of organizations and then the admin tries to change the state to any state other than `CourseCreator.GRANTED`. For example, this will happen when the request is denied after granting it.

farhaanbukhsh replied:
I have removed this validation and instead, when any of the statuses is changed from Granted to any other I remove all the orgs associated with it.

lgp171188 commented on cms/djangoapps/course_creators/admin.py:
```diff
@@ -30,26 +33,53 @@ def get_email(obj):
+
+    def clean(self):
+        """
+        Running validation of the form. Validation to stop user from entering
+        false data.
```
```suggestion
        Validate the 'state', 'orgs' and 'all_orgs' field before saving.
```

lgp171188 commented on common/djangoapps/student/roles.py:
```diff
@@ -245,6 +245,33 @@ class OrgRole(RoleBase):
+        from common.djangoapps.student.models import CourseAccessRole  # lint-amnesty, pylint: disable=redefined-outer-name, reimported
+        org_role_to_delete = []
+        orgs_role_to_create = []
+        existing_org_role = set(CourseAccessRole.objects.filter(user=user, role=self._role_name)
```
```suggestion
        existing_org_roles = set(CourseAccessRole.objects.filter(user=user, role=self._role_name)
```

farhaanbukhsh replied:
IMHO if the org roles are empty, it would not even reach the `if org_role_to_delete` block since it will be an empty set, so I think this will be a redundant check.

lgp171188 commented on common/djangoapps/student/roles.py:
```diff
@@ -245,6 +245,33 @@ class OrgRole(RoleBase):
     def __repr__(self):
         return f'<{self.__class__.__name__}>'
 
+    def update_org_course_role(self, user, orgs):
```
@farhaanbukhsh, should this be named `update_org_course_role` even though the docstring for the role says `A named role in a particular org independent of course`?

janimo reviewed (commented):
I've commented on two minor issues and I'll continue with testing.

It would be helpful to squash a few commits that change the same file if possible, but at least eliminate the initial commit and it's revert, to save time for those who start to review commit by commit :)

janimo reviewed (approved):
:+1:

- [x] I tested this: steps 1 to 16
- [x] I read through the code
- [ ] I checked for accessibility issues
- [ ] Includes documentation
- [ ] I made sure any change in configuration variables is reflected in the corresponding client's `configuration-secure` repository.

This looks ok, there are a couple minor things I found a bit confusing initially:

1) the user error message does not distinguish between user not having access to a certain groups and course creation being entirely disabled
"
Studio's having trouble saving your work

User does not have the permission to create courses in this organization or course creation is disabled
"
2) In the studio admin at admin/course_creators/coursecreator when ticking All organizations I was expecting it to ignore whatever is in the available/chosen panes for orgs, since All organizations supersedes those specific org settings. But if there were values left in Choses orgs it gives the same error as when nothing is selected at all

"The role can be granted either to all organizations or to specific organizations but not both."
and one has to remove those from the chosen pane. Not a big deal, it was just confusing the first few times :)

farhaanbukhsh commented:
@natabene  @kdmccormick  this up for review again :)

kdmccormick commented:
Thanks for your patience @farhaanbukhsh . Taking a look today.

kdmccormick commented on common/djangoapps/student/roles.py:
```diff
@@ -218,6 +218,33 @@ def users_with_role(self):
         )
         return entries
 
+    def update_org_role(self, user, orgs):
```
This method seems out of place.

An instance of `RoleBase` is meant to encapsulate a role name, an optional organization, and an optional course. The operations available on a `RoleBase` instance are adding users, removing users, and checking whether a user has the role.

This method does something completely else. It assumes that the `org` and `course_key` are None, and then mutates the roles across a range of organizations. This creates some weird behaviors. For example, if you executed this:
```python
OrgContentCreatorRole(org='org1').update_org_role(user, ['org2'])`
```
then the user would be assigned to just `org2`.

I recommend changing `update_org_role` in in `auth.py` to something like:
```python
def update_user_role_organizations(caller, role_class, user, *orgs):
    """
    ... docstring
    """
    _check_caller_authority(caller, role)
    for org in orgs:
        role_class(org=org).add_user(user)
```
letting the existing `RoleBase` functions take care of the CourseAccessRole table and caching concerns.

kdmccormick commented on cms/djangoapps/contentstore/views/course.py:
```diff
@@ -849,16 +850,18 @@ def _create_or_rerun_course(request):
+                             auth.user_has_role(request.user, OrgContentCreatorRole(org=org)))
+
+        if not is_course_creator:
+            raise PermissionDenied()
```
Ack, looks like we check course creator access in two other places as well:
* `_get_course_creator_status` in this same file
* `get_library_creator_status` in library.py

Both of those locations look at the `CourseCreator` table directly instead of looking at the auth roles. You will need to update both locations those to include the new organization-specificity logic.

farhaanbukhsh replied:
> _get_course_creator_status in this same file

This function is mostly responsible to view the list of content and IMHO this is okay to keep as is. I don't see any issues in this function, do let me know your thought about it.

farhaanbukhsh replied:
> get_library_creator_status

This function also looks at a higher level to see if the user has the permission to create a library, I think we should filter the Org part while creating the library.

kdmccormick replied:
>  _get_course_creator_status in this same file

> This function is mostly responsible to view the list of content and IMHO this is okay to keep as is. I don't see any issues in this function, do let me know your thought about it

Hm, I do think the function should be made organization-aware. It would be incongruent if `library.get_library_creator_status` function knew about organizations, but `course._get_course_creator_status` did not take organizations into account.

Plus, that should enable you to keep the implementation logic:
```
(
    auth.user_has_role(user, CourseCreatorRole()) or
    auth.user_has_role(user, OrgContentCreatorRole(org=org)
)
```
within a single utility function, instead of repeating it in multiple places.

kdmccormick reviewed (changes requested):
Great work so far @farhaanbukhsh . You are working against some tough legacy code here.

I tested your change out on my devstack, and it worked great!

I have some specific comments about your implementation below. I also have a more general consideration:

As I read through your PR and re-familiarized myself with the course-creator code, I actually cannot find any reason for `CourseCreatorRole` and `OrgContentCreatorRole` to exist. Studio access-checking code can look directly at the `CourseCreator` table (and some spots in `library.py` and `course.py` do exactly that!). I am not convinced that LMS needs to know anything about the course creator system. 

If you wanted to take a stab at simplifying all this by removing `CourseCreatorRole` altogether and just querying the `CourseCreator` directly instead, I would welcome that. However, I realize you have put a lot of time and effort into this implementation, so if you want to keep going with `CourseCreatorRole`/`OrgContentCreatorRole` then that is perfectly fine with me as well.

farhaanbukhsh commented:
@kdmccormick thanks for the review :), will address them and ping you :)

farhaanbukhsh commented on cms/djangoapps/course_creators/models.py:
```diff
@@ -2,17 +2,17 @@
 from django.utils import timezone
 from django.utils.encoding import python_2_unicode_compatible
 from django.utils.translation import ugettext_lazy as _
+from organizations.models import Organization  # lint-amnesty, pylint: disable=wrong-import-order
```
Makes sense :)

farhaanbukhsh commented on common/djangoapps/student/auth.py:
```diff
@@ -90,6 +92,9 @@ def get_user_permissions(user, course_key, org=None):
         return all_perms
     if course_key and user_has_role(user, CourseInstructorRole(course_key)):
         return all_perms
+    # User has all permission for libraries and courses if user belongs to content creator role.
+    if OrgContentCreatorRole(org=org).has_user(user):
+        return all_perms
```
Because when CourseContentGroup is granted user is given a `Staff` and `Instructor` role. But for the specific organization we give the user the `OrgContentCreatorRole` so we had to introduce the new role here. Hope I am able to clarify this.

kdmccormick replied:
> Why do we need to do this for OrgContentCreatorRole if we don't do it for CourseCreatorRole?

> Because when CourseContentGroup is granted user is given a Staff and Instructor role. But for the specific organization we give the user the OrgContentCreatorRole so we had to introduce the new role here. Hope I am able to clarify this.

@farhaanbukhsh Sorry, I don't fully understand. Are you saying that when `CourseCreatorRole` is granted to someone, they are given the `CourseStaff` and `CourseInstructor` access roles? How would that work when course-creator is a global role, and staff/instructor are course-specific roles?

farhaanbukhsh replied:
@kdmccormick `CourseStaff` and `CourseInstructor` access roles are assigned to the person who creates the role.

The check here is necessary because `OrgContentCreatorRole` deals with the creation of course so we need to raise error when user not belonging to the org. is creating a course.

While CourseCreatorRole is required to access `course creation` page and not check anything to do with course access. Hence we require the check for organization.

Hope I am able to clarify this.

kdmccormick replied:
@farhaanbukhsh This is still unclear to me. I cannot find any code paths that provide `org` as an argument to `get_user_permission`. Could you provide to me a specific scenario in which having this check here is necessary?

farhaanbukhsh replied:
`org` is not passed as a parameter to `get_user_permission` but deduced from course key [here](https://github.com/edx/edx-platform/blob/master/common/djangoapps/student/auth.py#L80), hence there is no need for explicitly passing org.



> Could you provide to me a specific scenario in which having this check here is necessary?

This is a very crucial check where write access for a user is checked for creating courses. You can have a look at this [test case](https://github.com/edx/edx-platform/pull/26616/files#diff-87512b8ef163801a02115105c1fc230d197ca6a00e4185cef51f2181212dc399R301) where write access is based on the role given to the user.

This check is indirectly used in a lot of places but the code path that touches this function in this PR starts [here](https://github.com/edx/edx-platform/blob/master/cms/djangoapps/contentstore/views/library.py#L283) and  [here](https://github.com/edx/edx-platform/blob/master/cms/djangoapps/contentstore/views/course.py#L283)

This is a non-exhaustive list.

Hope it's clearer now.

kdmccormick replied:
@farhaanbukhsh 
> This check is indirectly used in a lot of places but the code path that touches this function in this PR starts here and here

I tried removing this check, and found that the course creator still has instructor+staff access to the course. This is because we explicitly [grant instructor access](https://github.com/edx/edx-platform/blob/d5bacad86f9e37a04d58b26335fcd1967ffd99b8/cms/djangoapps/contentstore/views/course.py#L964) to a user when they create a course, rerun a course, or create a library.

farhaanbukhsh replied:
Tracing back the change I feel you are right, this check is not necessary now. It's possibly a left over from some approach that didn't work. I am sorry. :(

kdmccormick replied:
No sweat! Glad we worked through it.

farhaanbukhsh commented:
Hey @kdmccormick I have addressed your comments and left a reply for the questions. Thanks a lot for reviewing this PR so thoroughly. I took the roles way and not to tap in the CourseCreator entry because I find that solution to be more elegant. Let me know what you think about it. :smile: 

This is PR is up for review now.

kdmccormick commented:
Thanks @farhaanbukhsh ; I'll take a look when I can!

kdmccormick commented on cms/djangoapps/course_creators/admin.py:
```diff
@@ -30,26 +33,59 @@ def get_email(obj):
 get_email.short_description = 'email'
 
 
+class CourseCreatorForm(forms.ModelForm):
```
Could you include a screenshot of the modified admin page, as well as a screenshot of every possible new view or error message that a Studio user might see? I'm going to have a my product manager give this a look-over too.

farhaanbukhsh replied:
Here are the set of screenshots for all errors


**When all_org is not selected and no org is added**

![image](https://user-images.githubusercontent.com/7670449/128209710-11b91724-e973-400c-9eb6-4488fbd25559.png)

**When all_org and specific org both are selected**

![image](https://user-images.githubusercontent.com/7670449/128210135-a13128ad-a60e-4763-b925-93db9e562956.png)

**Studio Error when we put wrong org name**

![image](https://user-images.githubusercontent.com/7670449/128211264-ff9e3265-c462-4653-a687-c922c831e409.png)

kdmccormick replied:
@farhaanbukhsh Excellent. Can you move these to the PR description, and update description text as well? Try to separate the product change description from the engineering implementation details and testing instructions.

I will be showing this to my product manager soon. The clearer it is to her what the change is and how it'll affect the Studio product, the quicker we can get this review wrapped up and the PR merged 😃

farhaanbukhsh replied:
@kdmccormick  Thanks a bunch for looking at this 😄 . I have added a `product change description` in the description and have tried to explain the changes from a `product` POV. Please let me know if it's okay or should we make some changes.

Hope this PR gets to completions soon. 🤞🏾

kdmccormick replied:
@farhaanbukhsh Perfect, thank you.

gabor-boros commented:
Hey @farhaanbukhsh,
When you have time, could you please take a look at the comments above?

farhaanbukhsh commented:
@kdmccormick  Please have a look at the PR, thanks a lot for your patience. 😃

kdmccormick commented:
@farhaanbukhsh Thanks for the updates; I'll make another review pass soon!

farhaanbukhsh commented:
@kdmccormick  Thanks a bunch for looking at this 😄 . I have added a `product change description` in the description and have tried to explain the changes from a `product` POV. Please let me know if it's okay or should we make some changes.

Hope this PR gets to completions soon. 🤞🏾

cc-okoro commented:
@kdmccormick - This PR passed product review.

kdmccormick commented:
Thanks @cc-okoro !

@farhaanbukhsh Please rebase. I'll do final pass soon and we can get this into master 😄

farhaanbukhsh commented:
> @farhaanbukhsh Please rebase. I'll do final pass soon and we can get this into master smile

@kdmccormick  I have rebased the PR :) let me know if everything is okay. Thanks for your patience.

kdmccormick commented on cms/djangoapps/course_creators/admin.py:
```diff
@@ -75,22 +111,31 @@ def save_model(self, request, obj, form, change):
+    def save_related(self, request, form, formsets, change):
+        super().save_related(request, form, formsets, change)
+        state = form.instance.state
+        if state != CourseCreator.GRANTED:
+            form.instance.organizations.clear()
```
This functionality--specifically, that the many-to-many clears itself if `GRANTED` is not chosen--could be a bit confusing to users of the admin panel. It is not typical that a form silently clears itself based on the contents on another piece of the form. Is there a reason we need to do this?

kdmccormick replied:
This can be disregarded ^

kdmccormick reviewed (commented):
I was testing this in devstack this afternoon; just one comment so far. I'm hoping to finish my review pass tomorrow morning.

kdmccormick commented on cms/djangoapps/course_creators/admin.py:
```diff
@@ -30,26 +33,59 @@ def get_email(obj):
+        if is_state_granted:
+            if is_all_org_selected_with_orgs:
+                raise ValidationError(
+                    "The role can be granted either to ALL Organizations or to "
+                    "specific organizations but not both."
```
```suggestion
                    "The role can be granted either to ALL organizations or to "
                    "specific organizations but not both."
```
nit: inconsistent capitalization of "organizations"

kdmccormick commented on common/djangoapps/student/tests/test_roles.py:
```diff
@@ -142,6 +141,18 @@ def test_add_users_doesnt_add_duplicate_entry(self):
+        Test get_org_for_user
+        """
+        role = OrgContentCreatorRole(org=self.orgs[0])
+        assert len(list(role.get_org_for_user(self.student))) == 0
```
nit: No need to call `list` here, since your `get_org_for_user` function already returns a list.

kdmccormick commented on common/djangoapps/student/roles.py:
```diff
@@ -218,6 +218,12 @@ def users_with_role(self):
         )
         return entries
 
+    def get_org_for_user(self, user):
```
```suggestion
    def get_orgs_for_user(self, user):
```
This returns multiple short names, so please make the function name plural.

kdmccormick reviewed (commented):
I manually tested this again, and it works seamlessly. Nice job.

You can disregard [my concern about clearing organization m2m](https://github.com/edx/edx-platform/pull/26616#discussion_r704784066) -- after reviewing your implementation again, it seems like you chose the approach with the fewest potential edge cases, and I think the slight UI oddity of the m2m being cleared is acceptable.

Just some more code cleanup requests.

kdmccormick reviewed (changes requested):
(just marking this as "Changes requested" for the Jira automation; see my review above)

farhaanbukhsh commented on cms/djangoapps/contentstore/views/course.py:
```diff
@@ -1839,7 +1852,11 @@ def _get_course_creator_status(user):
         course_creator_status = 'disallowed_for_this_site'
     elif settings.FEATURES.get('ENABLE_CREATOR_GROUP', False):
         course_creator_status = get_course_creator_status(user)
-        if course_creator_status is None:
+        has_course_creator_role = True
+        if org:
+            has_course_creator_role = is_content_creator(user, org)
+
+        if course_creator_status is None or not has_course_creator_role:
```
Ahh! I should have been clearer with my persuasion :)

farhaanbukhsh commented:
@kdmccormick  This is ready for the second pass, thanks for the patience. :)

kdmccormick reviewed (approved):
Thank you for sticking with this PR and providing prompt responses even when I took I long time between rounds of feedback! I predict that some operators in the community will find the feature very useful.

kdmccormick commented:
@farhaanbukhsh I am using your Product description as the squashed commit message body because I find it to be a great explanation of the change.

kdmccormick commented:
@farhaanbukhsh I was about to Squash+Merge, and realized it'd be good to let you rebase and let tests run one one more time before doing so. If you do that, I will merge at my earliest convenience.

Below is the squashed commit message I had ready to go, but if you'd rather squash and prepare your own commit message, that's works too!
```
feat: grant course/library creation rights by organization (#26616)

Current State (before this commit):

  Studio, as of today doesn't have a way to restrict a user to
  create a course in a particular organization. What Studio
  provides right now is a CourseCreator permission which gives
  an Admin the power to grant a user the permission to create
  a course.

  For example: If the Admin has given a user Spiderman the
  permission to create courses, Spiderman can now create courses
  in any organization i.e Marvel as well as DC.
  There is no way to restrict Spiderman from creating courses
  under DC.

Purpose of this commit:

  The changes done here gives Admin the ability to restrict a
  user on an Organization level from creating courses via the
  Course Creators section of the Studio Django administration
  panel.

  For example: Now, the Admin can give the user Spiderman the
  privilege of creating courses only under Marvel organization.
  The moment Spiderman tries to create a course under some
  other organization(i.e DC), Studio will show an error message.

  This change is available to all Studio instances that
  enable the FEATURES['ENABLE_CREATOR_GROUP'] flag, but will
  not affect any instances that choose not to use it.

BB-3622
```

farhaanbukhsh commented:
@kdmccormick I have rebased the PR, feel free to use the commit message while doing a Squash+Merge. :) Thanks for the amazing reviews and a crazy lot of patience. :)

kdmccormick commented:
I added notes about this new feature to the [Maple release notes page](https://openedx.atlassian.net/wiki/spaces/COMM/pages/2603816700/Maple).
